lib/chef/provisioning/convergence_strategy/install_cached.rb

```diff
@@ -42,14 +42,26 @@ def initialize(convergence_options, config)
       def setup_convergence(action_handler, machine)
         super
 
-        # Install chef-client.  TODO check and update version if not latest / not desired
-        if machine.execute_always('chef-client -v').exitstatus != 0
-          platform, platform_version, machine_architecture = machine.detect_os(action_handler)
-          package_file = download_package_for_platform(action_handler, machine, platform, platform_version, machine_architecture)
-          remote_package_file = "#{@tmp_dir}/#{File.basename(package_file)}"
-          machine.upload_file(action_handler, package_file, remote_package_file)
-          install_package(action_handler, machine, remote_package_file)
+        # Check for existing chef client.
+        version = machine.execute_always('chef-client -v')
+
+        # Don't do install/upgrade if a chef client exists and
+        # no chef version is defined by user configs or
+        # the chef client's version already matches user config
+        if version.exitstatus == 0
+          if !chef_version
+            return
+          elsif version.stdout.strip =~ /Chef: #{chef_version}([^0-9]|$)/
```

Comment:
I think the only way to prevent updating the resource every run is to specify the exact version of chef that is desired, or for us to duplicate logic from the install.sh script.  If the user provides `12.1` as the desired chef version, `12.1.0` is the currently installed version and `12.1.1` was just released, what should we do?  Either we call the install.sh script and let it handle updating, or we duplicate the logic in install.sh so we can know whether we _should_ update.

I'm now realizing we may have to duplicate that logic regardless.  install.sh is only 1 of the 3 convergence strategies here.  The other 2 won't be able to answer the above question unless we access the omnitruck API.

My vote is that we change it to exact matching for all 3 convergence strategies.  This will make it idempotent for users who specify the full version of chef they want to install.  Users who specify a partial version will get the existing behavior, so we should file a bug to come back and fix that.  Sound good?